app/src/main/admin/actions/modellingGroupsActionCreators.ts

```diff
@@ -0,0 +1,100 @@
+            const groups: ModellingGroup[] = await (new ModellingGroupsService(dispatch, getState)).getAllGroups();
+            dispatch({
+                type: ModellingGroupTypes.ADMIN_GROUPS_FETCHED,
+                data: groups
+            } as AdminGroupsFetched );
+        }
+    },
+
+    getGroupDetails(groupId: string) {
+        return async (dispatch: Dispatch<AdminAppState>, getState: () => AdminAppState) => {
+            const groupDetails: ModellingGroupDetails = await (new ModellingGroupsService(dispatch, getState)).getGroupDetails(groupId);
+            dispatch({
+                type: ModellingGroupTypes.ADMIN_GROUP_DETAILS_FETCHED,
+                data: groupDetails
+            } as SetAdminGroupDetails);
+        }
+    },
+
+    setCurrentGroup(groupId: string) {
+        return (dispatch: Dispatch<AdminAppState>, getState: () => AdminAppState) => {
+            const groups = getState().groups.groups;
+            const currentGroup = groups.find(group => group.id === groupId);
+            dispatch({
+                type: ModellingGroupTypes.SET_CURRENT_ADMIN_GROUP,
+                data: currentGroup ? currentGroup : null
+            } as SetCurrentAdminGroup);
+        }
+    },
+
+    addUserToGroup(groupId: string, username: string) {
+        return async (dispatch: Dispatch<AdminAppState>, getState: () => AdminAppState) => {
+            const result = await (new ModellingGroupsService(dispatch, getState))
+                .addMember(groupId, username);
+            if (result === "OK") {
+                dispatch(this.clearCacheForGroupDetails(groupId));
+                await dispatch(this.getGroupDetails(groupId));
+                dispatch(this.setCurrentGroupMembers());
+            }
+        }
+    },
+
+    removeUserFromGroup(groupId: string, username: string) {
+        return async (dispatch: Dispatch<AdminAppState>, getState: () => AdminAppState) => {
+            const result = await (new ModellingGroupsService(dispatch, getState))
+                .removeMember(groupId, username);
+            if (result === "OK") {
+                dispatch(this.clearCacheForGroupDetails(groupId));
+                await dispatch(this.getGroupDetails(groupId));
+                dispatch(this.setCurrentGroupMembers());
+            }
+        }
+    },
+
+    clearCacheForGroupDetails(groupId: string,) {
+        return (dispatch: Dispatch<ContribAppState>, getState: () => ContribAppState) => {
+            (new ModellingGroupsService(dispatch, getState)).clearCacheForGroupDetails(groupId);
+        }
+    },
+
+    setCurrentGroupMembers() {
+        return (dispatch: Dispatch<AdminAppState>, getState: () => AdminAppState) => {
+            const currentGroupDetails = getState().groups.currentGroupDetails;
+            const allUsers = getState().users.users;
+            if (!currentGroupDetails || !allUsers.length) {
```

Comment:
So if `currentGroupDetails` exists or `allUsers` exists, this won't dispatch; why do we check this?

Reply:
it maps users to group details members usernames list, if something doesn't exists, it cannot be mapped, so dispatch empty

Reply:
before that you had this mapping in many places in components, i moved that mapping logic out to actions and reducers, so that we can map it once and use where we need

Reply:
don't quite understand what you mean. Local variable? Where?

Reply:
Sorry, should have been more clear:

```
            const currentGroupDetails = getState().groups.currentGroupDetails;
            const allUsers = getState().users.users;
            const currentGroupMembers = [];

            if (currentGroupDetails && allUsers.length) {
                currentGroupMembers = compact(currentGroupDetails.members
                .map(memberUsername => allUsers.find(user => user.username === memberUsername))
            }

             dispatch({
                type: ModellingGroupTypes.SET_CURRENT_ADMIN_GROUP_MEMBERS,
                data: currentGroupMembers
           } as SetCurrentAdminGroupMembers);
```

Reply:
Yes it is a good idea, to chain actions, because it will never will be called independently anyway. Always set members after details fetch. I actually has this idea in mind from the start, then just decided by some reason to do it independently.